### ISSUE TITLE
Provide `StateStorage` abstraction

### ISSUE THREAD

**Issue by klimov-paul** [yiisoft/yii2-authclient#121]
Methods `BaseOAuth::setState()` and `BaseOAuth::getState()` should be improved.
It should be a separated configurable component, which can handle state saving.
Such component can be adjusted in case default one is not suitable by some reason.

**Comment by samdark** [comment]
I guess state itself should be an object, right?

**Comment by samdark** [comment]
```php
interface StateStorageInterface
{
    public function set($key, $value);
    public function get($key);
    public function remove($key);
}
```

**Comment by samdark** [comment]
Right?

**Comment by klimov-paul** [comment]
Yes.

**Comment by klimov-paul** [comment]
Resolved by commit b04a83d
